discovery_engine_core/providers/src/models/content.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Error, NewscatcherArticle};
+use chrono::NaiveDateTime;
+use serde::{Deserialize, Serialize};
+use std::ops::Deref;
+use url::Url;
+
+/// A helper type used to ensure that, within the [`GenericArticle`] struct,
+/// we never use a URL which does not have a domain, such as `file:///foo/bar`.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct UrlWithDomain(Url);
+
+impl Deref for UrlWithDomain {
+    type Target = Url;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl UrlWithDomain {
+    pub fn new(url: Url) -> Option<Self> {
+        if url.domain().is_none() {
+            None
+        } else {
+            Some(UrlWithDomain(url))
+        }
+    }
+
+    #[allow(clippy::missing_panics_doc)]
+    pub fn domain(&self) -> &str {
+        self.0.domain().unwrap(/* constructor makes sure we have a domain */)
+    }
+
+    pub fn inner(&self) -> Url {
+        self.0.clone()
+    }
+}
+
+impl TryFrom<&str> for UrlWithDomain {
+    type Error = ();
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Ok(UrlWithDomain::new(Url::parse(value).unwrap()).unwrap())
+    }
+}
+
+/// Represents a news that is delivered by an external content API.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GenericArticle {
+    pub title: String,
+    pub snippet: String,
+    pub url: UrlWithDomain,
+    pub date_published: NaiveDateTime,
+    pub country: String,
+    pub language: String,
+    pub topic: String,
+    pub image: Option<Url>,
+
+    /// Private so that we can centrally control the default
+    /// value of `rank`
```

Comment:
I felt that this would be easier than trying to synchronise the default value across the various places where we require `rank` to exist. It's true that this prevents creation outside of `providers`, but if we really needed to do that we could also implement a constructor for it, I suppose (?) I don't feel strongly about this, though, tbh.

Reply:
Would you be okay with making the field `pub`, but calling it `raw_rank`?

Reply:
we could also do 
```rust 
        pub struct Rank(pub u32);

        impl Default for Rank {
            fn default() -> Self {
                Self(1)
            }
        }

        #[derive(Default)]
        pub struct Article {
            pub title: String,
            pub rank: Option<Rank>,
        }

        let article = Article {
            title: "title".to_string(),
            ..Article::default()
        };

        assert!(article.rank.unwrap_or_default().0 == 1);
```

Reply:
Does this look good?

Reply:
lgtm 👍